### ISSUE TITLE
Change of default attribute/metadata type when value changes type

### ISSUE THREAD

**Issue by kzangeli** [telefonicaid/fiware-orion#2451]
This issue is a problem related to issue #2223.

Imagine the following scenario:

1. Entity E with attribute A is created. A has no given type and its value is 14.
- A is given the default type `Number`

2. A is modified using /v2/entities/attrs/A/value, changing the value of A to `false`.
- A already has a type, the broker doesn't know that A's type came from a default value.

This scenario was OK when we had `none` as unique default type.
Now, it is a problem.
Only way to fix this (in step 02, changing the type of A to ' Boolean` as no type has been given) is to have the info about *typeGiven* in mongo.

Until this issue is thought over and solved, the following functests work in the way that the broker works right now, the type is not modified when changing an attribute from 14 to false. The type that came from default type `Number` survives the change of A's value to `true`.

The following functests are affected:
* 0999_update_single_attr_value_by_eid/update_single_attr_by_eid.test
* 0999_update_single_attr_value_by_eid/update_single_attr_value_by_eid_fake_field_cases.test
* 1185_udpate_change_value_type/update_change_value_type.test
* 1188_empty_attribute_value/empty_attribute_value.test
* 1259_new_json_for_v2/put_attribute_value.test

**Comment by jmcanterafonseca** [comment]
my proposal to fix this is that for those attributes for which the user didn't specify a type, to make the attribute type a slave of the attribute value, i.e. every time the attribute value changes, the attribute type changes accordingly, even if the user only changed the attribute value.

Implementation of this may require on the fly generation of attribute types ...

**Comment by fgalan** [comment]
Considering a generic REST API, if we have a resource identified by `/path/to/resource/R` with 3 resource types, e.g. foo, bar and mu) then the PUT on a given sub-resource shouldn't touch the others. I mean:

* PUT /path/to/resource/R/foo should modify foo, but not bar or mu
* PUT /path/to/resource/R/bar should modify bar, but not foo or mu
* PUT /path/to/resource/R/mu should modify mu, but not foo or bar

Applying this generic rule to our NGSIv2 case:

* PUT /v2/entities/E/attrs/A/value should modify value, but not type or metadata
* PUT /v2/entities/E/attrs/A/type should modify type, but not value or metadata
* PUT /v2/entities/E/attrs/A/metadata should modify metadata, but not value or type

(Actually, the second and third operations doesn't currently exist in the NGSIv2 but I'm using then to expose my point. Note that they could be added in a future evolution of NGSIv2).

Taking into account the above, "binding" type as an slave of value would mean to break REST general principles.

On the other hand, introducing an extra status for attribute about if the type was provided or not by the user would make semantic more complex, thus making the API more difficult to use. In some sense, doing so would go against the whole idea of type defaulting (which should be as transparent as if the user have provided the type in an explicit way). In addition, it would involve to introduce an additional state at MongoDB level, which is complex.

Thus, I think the scenario exposed by @kzangeli is the right behaviour. I think is a small price to pay, taking into account other options would be worse. If the user want to update an attribute value with type defaulting he/she should use the PUT /v2/entities/E/attrs/A operation, in the following way:

```
PUT /v2/entities/E/attrs/A
{ "value": false }
```

which, applying partial entity representation rule *Attribute/metadata type may be omitted in requests. When omitted in attribute/metadata creation or in update operations, a default is used for the type depending on the value*, will change attribute value to boolean false and attribute type to `Boolean` (it also changes metadata vector to `[]` but that's another history... :) )

In conclusion, I think no change regarding this is needed. Hoaver, this issue probalby raises the question about if the current set of operations is to limitted, e.g. how to change type without changing value or metadata? How to create new metadata without touching existing ones?

**Comment by kzangeli** [comment]
While Fermins point is perfectly valid, I personally would prefer to either **rollback the entire idea** of different default types, and use **none** for everything,

  **OR**

  add a field in the database to record that an attribute/metadata should carry the default type.

The way it is right now (PR still to be merged into develop), an attribute could (yes, definitely a corner-case) end up looking like this:

    "Attr1": {
        "type": Boolean",
        "value": 14
    }

And ... that is just **WEIRD** (IMHO).

It was better before: "type": "none".
After all, these default types are **only to state what is  already obvious**: 14 is Number, false is Boolean, etc.

**Comment by iariasleon** [comment]
Only a comment, all changes and verifications of types, penalize in performance, and it is something to we should think too.

**Comment by fgalan** [comment]
There is another option: if PUT /v2/entities/E/attrs/A/value would produce a change in the JSON datatype, return an error. E.g. if the attribue is number 14, trying to change to boolean false will return an error. It doesn't have the problem of implementation complexity (doesn't need additional fields in the DB) and doesn't make semantics more complex.

I don't support it as it would make the usage of the API more complex and introduce an "asymetry" in the operations to change attributes (other operations always allow to change JSON datatype), but I put it on the table.

**Comment by fgalan** [comment]
There is another option: if PUT /v2/entities/E/attrs/A/value would produce a change in the JSON datatype, return an error. E.g. if the attribue is number 14, trying to change to boolean false will return an error. It doesn't have the problem of implementation complexity (doesn't need additional fields in the DB) and doesn't make semantics more complex.

I don't support it as it would make the usage of the API more complex and introduce an "asymetry" in the operations to change attributes (other operations always allow to change JSON datatype), but I put it on the table.

**Comment by kzangeli** [comment]
Strongly typed attributes ...
I wouldn't mind it. It's weird that an attribute is created as a Boolean and then updated to be a string instead. Very weird (which is why this whole thing is so corner-case).

**Comment by jmcanterafonseca** [comment]
I think we may have this problem even if there is not default datatyping. Imagine I create an attribute

{
   "fancyAttr": {
      "value": 33,
      "type": "Number"
   }
}

and then update value to be `false`. Automatically the potential inconsistency is there. 

However in the JSON / JS world there is type coercion so I think it is not so bad the situation of type and value not matching directly, because type coercion could automatically solve the potential inconsistency and that could be left in hands of the application. 

Having said that I think a good approach would be do nothing. The idea of validation and raising error although it can be appealing might open a can worms, specially with data types like dates , times, etc. That for consistency would force us to validate them and that's now for the moment in scope I think

**Comment by jmcanterafonseca** [comment]
just to note that PATCH operation should be used for changing metadata or type without changing value, that's issue number #1786

**Comment by kzangeli** [comment]
About the example with "fancyAttr":
There is a difference here. A pretty important one.
In the example with "fancyAttr", the **user** decides to give it the type "Number" and after that changing the value to a Boolean, without also changing its type.
If the user wants to do that, it is his problem.

The original case is very different.
The **broker** invents the type for the user, and after that it goes wrong.
This is a failure in the broker, imposing stuff on the user, that the user probably doesn't want.

Very different cases indeed

**Comment by fgalan** [comment]
After discussion, we will opted at the end to leave as it is now (or, more precisely, to leave as it would be after merging PR #2469)
